packages/ra-core/src/form/formMiddleware.js

```diff
@@ -0,0 +1,16 @@
+import { LOCATION_CHANGE } from 'react-router-redux';
+import { beforeLocationChange } from '../actions/formActions';
+
+const formMiddleware = () => next => action => {
```

Comment:
I and other developers will be unable to understand why you're doing that. This file requires many, many comments to describe the issue and the solution.